### PR TITLE
Np 2500 query request with real response

### PR DESCRIPTION
**Pull request by svenngjeto**
Mocking call to NSD in testing, using real data. Request and client separated to own source files

### CODE REVIEW

File: src/requestgenerator.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+  createRequest (event) {
+    const params = event.queryStringParameters
+    let template
+    if (event.path === '/journal') {
+      template = queryJournalTemplate
+    } else {
+      template = queryPublisherTemplate
+    }
```
Comment (brinxmat) [inline src/requestgenerator.js:19]: I'd probably go with
```
    const template = event.path === '/journal' ? readTemplate('journalQueryTemplate.json') : readTemplate('queryPublisherTemplate.json')
    …
    const readTemplate = (path) => JSON.parse(fs.readFileSync(path).toString())

```

File: src/requestgenerator.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      filterValue = '%' + params.query + '%'
+    }
+    template.filter[0].selection.values[0] = filterValue
+    return template
+  }
+}
+
+const queryJournalTemplate = JSON.parse(fs.readFileSync('journalQueryTemplate.json').toString())
+
+const queryPublisherTemplate = JSON.parse(fs.readFileSync('queryPublisherTemplate.json').toString())
```
Comment (brinxmat) [inline src/requestgenerator.js:31]: See previous comment

File: src/requestgenerator.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+const fs = require('fs')
```
Comment (brinxmat) [inline src/requestgenerator.js:1]: I'd rename the file to match other .js file naming, probably `Request.js` as it's a class

File: src/queryPublisherTemplate.json [hunk 4]
```diff
@@ -0,0 +1,29 @@
+{
```
Comment (brinxmat) [inline src/queryPublisherTemplate.json:1]: The filename here should match other json resources and use snake case

File: src/journalQueryTemplate.json [hunk 5]
```diff
@@ -0,0 +1,33 @@
+{
```
Comment (brinxmat) [inline src/journalQueryTemplate.json:1]: The file naming here should match other JSON resources and use snake case

File: src/tests/unit/handler.test.js [hunk 6]
```diff
@@ -37,11 +74,11 @@ describe("Handler verifies route /journal; path '/journal', httpMethod.GET", ()
   ['/journal', '/publisher'].map(calledPath => (
```
Comment (brinxmat) [inline src/tests/unit/handler.test.js:74]: Does this test do anything anymore? The commented code implies not — what are the body expectations?

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: src/client.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+const axios = require('axios')
+
+const responseWithBody = (body) => {
+  return {
+    statusCode: 200,
+    headers: {
+      'Content-Type': 'application/json'
+    },
+    isBase64Encoded: false,
+    body: body === undefined ? '{}' : JSON.stringify(body)
```
[the remark this answers is not in the document]
Comment (svenngjeto) [reply src/client.js:12]: Reusing old function with body parameter with default value for old usage

File: src/client.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  }
+}
+
+async function performQuery (request) {
```
[the remark this answers is not in the document]
Comment (svenngjeto) [reply src/client.js:16]: Yes, of course!

File: src/requestgenerator.js (hunk 2 above)
Comment (svenngjeto) [reply src/requestgenerator.js:31]: Code optimized and removed

File: src/requestgenerator.js (hunk 1 above)
Comment (svenngjeto) [reply src/requestgenerator.js:19]: Yes, simplifies code

File: src/queryPublisherTemplate.json (hunk 4 above)
Comment (svenngjeto) [reply src/queryPublisherTemplate.json:1]: Same name convention as the other query template, at least

File: src/journalQueryTemplate.json (hunk 5 above)
Comment (svenngjeto) [reply src/journalQueryTemplate.json:1]: same naming parts as the other

File: src/tests/unit/handler.test.js (hunk 6 above)
Comment (svenngjeto) [reply src/tests/unit/handler.test.js:74]: The test still verifiest that the routes is recognized and returns 200 OK for those

File: src/Request.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+class Request {
+  constructor (event) {
+    this.request = this.createRequest(event)
+  }
+
+  getRequest () {
+    return this.request
+  }
+
+  createRequest (event) {
```
Comment (axthosarouris) [inline src/Request.js:12]: I suggest a refactoring that will made the code more readable

```
const MYSTERIOUS_CONSTANT_IN_FILTER_VALUE_IN_TEMPLATES = '%';

class Request {
  constructor (event) {
    this.request = this.createRequest(event)
  }

  getRequest () {
    return this.request
  }

  createRequest (event) {
    const params = event.queryStringParameters
    const template = this.isJournal(event) 
        ? this.readJournalTemplate() 
        : this.readPublisherTemplate()
    
    let filterValue = this.crateQueryFilterBasedOnQueryParameters(params);
    this.doSomeExtraMagic(template, filterValue);
    return template
  }

  doSomeExtraMagic(template, filterValue) {
    template.filter[0].selection.values[0] = filterValue
  }

  crateQueryFilterBasedOnQueryParameters(params) {
    let filterValue = MYSTERIOUS_CONSTANT_IN_FILTER_VALUE_IN_TEMPLATES
    if (this.queryHasNoParameters(params)) {
      filterValue = '%' + params.query + '%'
    }
    return filterValue;
  }

  queryHasNoParameters(params) {
    return params !== undefined && params.query !== undefined;
  }

  readPublisherTemplate() {
    return readTemplate('query_publisher_template.json');
  }

  readJournalTemplate() {
    return readTemplate('query_journal_template.json');
  }

  isJournal(event) {
    return event.path === '/journal';
  }
}
const readTemplate = (path) => JSON.parse(fs.readFileSync(path).toString())
```

File: src/Request.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    const params = event.queryStringParameters
+    const template = event.path === '/journal' ? readTemplate('query_journal_template.json') : readTemplate('query_publisher_template.json')
+    let filterValue = '%'
+    if (params !== undefined && params.query !== undefined) {
+      filterValue = '%' + params.query + '%'
+    }
+    template.filter[0].selection.values[0] = filterValue
```
Comment (axthosarouris) [inline src/Request.js:19]: It is  unclear to me what is happening here.

File: src/Request.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+  }
+
+  getRequest () {
+    return this.request
+  }
+
+  createRequest (event) {
+    const params = event.queryStringParameters
+    const template = event.path === '/journal' ? readTemplate('query_journal_template.json') : readTemplate('query_publisher_template.json')
+    let filterValue = '%'
```
Comment (axthosarouris) [inline src/Request.js:15]: reafactor the '%' to a constant with a name describing its meaning

File: src/tests/unit/handler.test.js [hunk 12]
```diff
@@ -1,14 +1,51 @@
 'use strict'
 
-const { handler } = require('../../handler')
+const handler = require('../../handler')
 const chai = require('chai')
 const expect = chai.expect
 const httpStatus = require('http-status-codes')
+const fs = require('fs')
+const nock = require('nock')
```
Comment (axthosarouris) [inline src/tests/unit/handler.test.js:8]: maybe better to name it httpServerMock? It was hard to google 'javascript nock', kept getting "do you main javascirpt mock"?

File: src/tests/unit/handler.test.js [hunk 13]
```diff
@@ -1,14 +1,51 @@
...
+}
+
+const hasValidQueryParameters = (requestBody) => {
+  const querySpec = [{ name: 'query', required: true }, { name: 'year', required: false }, { name: 'start', required: false }]
+  if (requestBody.queryStringKeys === undefined) return true
+  const queryStringKeys = Object.keys(requestBody.queryStringParameters)
+  const valid = isValidParameterName(queryStringKeys, querySpec) && hasRequiredParameters(queryStringKeys, querySpec)
+  console.log(`hasValidQueryParameters(${requestBody}) -> ${valid}`)
+  return valid
+}
```
Comment (axthosarouris) [inline src/tests/unit/handler.test.js:24]: Helper methods should be under the tests. I think having the same **general** codestlyle principles helps when we have a polyglot code base.

File: src/tests/unit/handler.test.js [hunk 14]
```diff
@@ -1,14 +1,51 @@
...
+  .reply(httpStatus.OK, publisherContent)
+
+// Fallback interceptor to catch invalid parameters
+nock('https://api.nsd.no/dbhapitjener/Tabeller/hentJSONTabellData')
+  .persist()
+  .post('')
+  .reply(httpStatus.INTERNAL_SERVER_ERROR, {})
 
 describe('Handler throws error when called', () => {
   it('verifies response is error 500 and  has Internal Server Error message', async function () {
```
Comment (axthosarouris) [inline src/tests/unit/handler.test.js:46]: when does this happen?

File: src/tests/unit/handler.test.js [hunk 15]
```diff
@@ -37,11 +74,10 @@ describe("Handler verifies route /journal; path '/journal', httpMethod.GET", ()
   ['/journal', '/publisher'].map(calledPath => (
     it(`GET ${calledPath} returns 200 OK and has empty body`, async function () {
       const httpMethod = 'GET'
-      const emptyBody = '{}'
       const event = { path: calledPath, httpMethod: httpMethod }
-      const response = await handler(event)
+      const response = await handler.handler(event)
       expect(response.statusCode).to.equal(httpStatus.OK)
-      expect(response.body).to.equal(emptyBody)
+      // expect(response.body).to.equal(emptyBody)
```
Comment (axthosarouris) [inline src/tests/unit/handler.test.js:80]: delete commented code

File: src/tests/unit/handler.test.js [hunk 16]
```diff
@@ -51,7 +87,7 @@ describe('Handler throws 405 when httpMethod is not GET', () => {
     it(`httpmMethod ${calledMethod} returns 405`, async function () {
```
Comment (axthosarouris) [inline src/tests/unit/handler.test.js:87]: http**mM**ethod

**Review by axthosarouris — CHANGES_REQUESTED** [review]
I dynamically typed languages it is important to focus on readablity. (Will continue after the meeting)

File: src/Request.js (hunk 9 above)
Comment (svenngjeto) [reply src/Request.js:12]: Renamed and refactored methods to clarify meaning

File: src/Request.js (hunk 10 above)
Comment (svenngjeto) [reply src/Request.js:19]: Given more descriptive name

File: src/Request.js (hunk 11 above)
Comment (svenngjeto) [reply src/Request.js:15]: Renamed to what it is , an SQL wilcard character

File: src/tests/unit/handler.test.js (hunk 13 above)
Comment (svenngjeto) [reply src/tests/unit/handler.test.js:24]: Helpers moved down to bottom

File: src/tests/unit/handler.test.js (hunk 14 above)
Comment (svenngjeto) [reply src/tests/unit/handler.test.js:46]: It don't.. Removed

File: src/tests/unit/handler.test.js (hunk 16 above)
Comment (svenngjeto) [reply src/tests/unit/handler.test.js:87]: Fatfingered m removed

**Review by axthosarouris — APPROVED** [review]

File: src/tests/unit/handler.test.js [hunk 17]
```diff
@@ -98,44 +110,80 @@ describe('Handler verifies queryStringParameters and returns 200 with empty body
...
+    it(`returns 200 OK for ${calledPath}`, async function () {
+      const queryStringParameters = { query: 'Alzheimers', year: 2020, start: 1 }
+      const event = { path: calledPath, httpMethod: 'GET', queryStringParameters: queryStringParameters }
+      const response = await handler.handler(event)
+      expect((await response).statusCode).to.equal(httpStatus.OK)
+    })
+  ))
+})
+
+describe('Handler returns response 200 OK and data when called', () => {
```
Comment (torarnet) [inline src/tests/unit/handler.test.js:164]: Are these two methods identical?

Comment (svenngjeto) [reply src/tests/unit/handler.test.js]: They are! Removed

File: src/Request.js [hunk 18]
```diff
@@ -0,0 +1,35 @@
+const fs = require('fs')
+
+class Request {
+  constructor (event) {
+    this.request = this.createRequest(event)
+  }
+
+  getRequest () {
```
Comment (brinxmat) [inline src/Request.js:8]: `get request () {…}`

See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: src/Request.js (hunk 18 above)
Comment (svenngjeto) [reply src/Request.js]: Added getter and setter, and test for coverage

File: src/tests/unit/Request.test.js [hunk 19]
```diff
@@ -0,0 +1,19 @@
+const requestGenerator = require('../../Request')
+const chai = require('chai')
+const expect = chai.expect
+
+describe('Testing getter and setter in request', () => {
```
Comment (axthosarouris) [inline src/tests/unit/Request.test.js:5]: I think the description of the tests should give a general description, and the "it" parts should have the structure 
"it(returns something when input is this and the status is that").

File: src/tests/unit/Request.test.js [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+  it('testing getter in request body and verify table_id', async function () {
```
Comment (axthosarouris) [inline src/tests/unit/Request.test.js:6]: I am not sure I can understand what is the intention with this test

File: src/publisher_response.json [hunk 21]
```diff
@@ -0,0 +1 @@
+[{"Original tittel":"T & AD Poyser","Forlag id":"25264","Url":"http://www.poyserbooks.co.uk/","Aktiv":"1"},{"Original tittel":"T.M.C. Asser Press","Forlag id":"19698","Url":"http://www.asser.nl","Aktiv":"1"},{"Original tittel":"Tallinn University Press","Forlag id":"24971","Url":"http://www.tlupress.com/?LangID=2&CatID=41","Aktiv":"1"},{"Original tittel":"Tamesis Books","Forlag id":"26824","Url":"https://boydellandbrewer.com/tamesis-books","Aktiv":"1"},{"Original tittel":"Tammi","Forlag id":"20259","Url":"http://www.tammi.fi","Aktiv":"1"},{"Original tittel":"Tampere University of Technology","Forlag id":"18630","Url":"http://www.tut.fi/","Aktiv":"1"},{"Original tittel":"Tampere University Press","Forlag id":"24781","Url":"http://www.uta.fi/laitokset/kirjasto/julkaisukeskus/tup/index_en.php","Aktiv":"1"},{"Original tittel":"Tapia House","Forlag id":"18912","Url":null,"Aktiv":"1"},{"Original tittel":"Tapir Akademisk Forlag","Forlag id":"18441","Url":"http://www.tapirforlag.no/","Aktiv":"0"},{"Original tittel":"TAPPI Press","Forlag id":"24927","Url":"http://www.tappi.org/Bookstore/Manuscript-Submissions.aspx","Aktiv":"1"}]
```
Comment (janbj) [inline src/publisher_response.json:1]: Test files in different folder?

File: src/tests/unit/Request.test.js [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+    const queryParameters = { path: '/journal', queryStringParameters: { query: 'query' } }
+    const nsdRequest = new requestGenerator.Request(queryParameters)
+    const request = nsdRequest.request
+    expect(request.tabell_id).to.equal(851)
+  })
+  it('testing setter in request ', async function () {
+    const queryParameters = { path: '/journal', queryStringParameters: { query: 'query' } }
+    const nsdRequest = new requestGenerator.Request(queryParameters)
+    nsdRequest.request = { tabell_id: 123 }
+    const requestBody2 = nsdRequest.request
```
Comment (axthosarouris) [inline src/tests/unit/Request.test.js:16]: why is this called requestBody2? what is the first requestBody?

File: src/query_journal_template.json [hunk 23]
```diff
@@ -0,0 +1,33 @@
+{
```
Comment (janbj) [inline src/query_journal_template.json:1]: Templates in different folder?

File: src/query_publisher_template.json [hunk 24]
```diff
@@ -0,0 +1,29 @@
+{
```
Comment (janbj) [inline src/query_publisher_template.json:1]: Queries in different folder?

File: src/tests/unit/handler.test.js [hunk 25]
```diff
@@ -1,14 +1,28 @@
...
 
-describe('Handler throws error when called', () => {
+httpServerMock('https://api.nsd.no', { reqheaders: { 'content-type': 'application/json;charset=utf-8' } })
+  .persist()
+  .post('/dbhapitjener/Tabeller/hentJSONTabellData', body => { return body.path === '/journal' && hasValidQueryParameters(body) })
+  .reply(httpStatus.OK, journalContent)
+
+httpServerMock('https://api.nsd.no', { reqheaders: { 'content-type': 'application/json;charset=utf-8' } })
+  .persist()
+  .post('/dbhapitjener/Tabeller/hentJSONTabellData', body => { return body.path === '/publisher' && hasValidQueryParameters(body) })
```
Comment (axthosarouris) [inline src/tests/unit/handler.test.js:19]: could this be a constant?

File: src/tests/unit/handler.test.js [hunk 26]
```diff
@@ -1,14 +1,28 @@
...
 const chai = require('chai')
 const expect = chai.expect
 const httpStatus = require('http-status-codes')
+const fs = require('fs')
+const httpServerMock = require('nock')
+const journalContent = fs.readFileSync('journal_response.json').toString()
+const publisherContent = fs.readFileSync('publisher_response.json').toString()
 
-describe('Handler throws error when called', () => {
+httpServerMock('https://api.nsd.no', { reqheaders: { 'content-type': 'application/json;charset=utf-8' } })
```
Comment (axthosarouris) [inline src/tests/unit/handler.test.js:12]: could this be a constant?

**Review by axthosarouris — CHANGES_REQUESTED** [review]

File: src/client.js [hunk 27]
```diff
@@ -0,0 +1,21 @@
+const channelRegistryUri = 'https://api.nsd.no/dbhapitjener/Tabeller/hentJSONTabellData'
+
+const axios = require('axios')
+
+const responseWithBody = (body) => {
```
Comment (janbj) [inline src/client.js:5]: Is it the right place to create a handler response in the client?

**Review by janbj — COMMENTED** [review]
Ver

File: src/tests/unit/handler.test.js (hunk 15 above)
Comment (svenngjeto) [reply src/tests/unit/handler.test.js]: deleted

File: src/tests/unit/Request.test.js (hunk 22 above)
Comment (svenngjeto) [reply src/tests/unit/Request.test.js:16]: It's just a Request's internal part...

File: src/tests/unit/handler.test.js (hunk 25 above)
Comment (svenngjeto) [reply src/tests/unit/handler.test.js:19]: Moved to constant

File: src/tests/unit/handler.test.js (hunk 26 above)
Comment (svenngjeto) [reply src/tests/unit/handler.test.js:12]: yes, make constant

File: src/tests/unit/handler.test.js (hunk 12 above)
Comment (svenngjeto) [reply src/tests/unit/handler.test.js]: Renamed

File: src/publisher_response.json (hunk 21 above)
Comment (svenngjeto) [reply src/publisher_response.json]: Test files moved to test...

**Review by brinxmat — APPROVED** [review]

File: src/tests/unit/Request.test.js [hunk 28]
```diff
@@ -0,0 +1,19 @@
+const requestGenerator = require('../../Request')
+const chai = require('chai')
+const expect = chai.expect
+
+describe('Testing getter and setter in Request class for code coverage', () => {
+  it('testing getter in request body and verify table_id', async function () {
+    const queryParameters = { path: '/journal', queryStringParameters: { query: 'query' } }
+    const nsdRequest = new requestGenerator.Request(queryParameters)
+    const request = nsdRequest.request
+    expect(request.tabell_id).to.equal(851)
```
Comment (axthosarouris) [inline src/tests/unit/Request.test.js:10]: rename to TABLE_ID_FOUND_IN_RESOURCE_FILE

**Review by axthosarouris — APPROVED** [review]

**Pull request merged**